packages/atomic/src/components/result-lists/result-list-common.tsx

```diff
@@ -0,0 +1,204 @@
+import {TemplateContent} from '../result-template/atomic-result-template/atomic-result-template';
+import {
+  getResultDisplayClasses,
+  ResultDisplayDensity,
+  ResultDisplayImageSize,
+  ResultDisplayLayout,
+} from '../atomic-result/atomic-result-display-options';
+import {AtomicFoldedResultList} from './atomic-folded-result-list/atomic-folded-result-list';
+import {AtomicResultList} from './atomic-result-list/atomic-result-list';
+
+export interface AtomicResultListBase {
+  bindings: Bindings;
+  host: HTMLElement;
+  templateHasError: boolean;
+  density: ResultDisplayDensity;
+  display?: ResultDisplayLayout;
+  imageSize?: ResultDisplayImageSize;
+  image: ResultDisplayImageSize;
+  listWrapperRef?: HTMLDivElement;
+  resultsPerPageState: ResultsPerPageState;
+
+  resultList: FoldedResultList | ResultList;
+  resultListState: FoldedResultListState | ResultListState;
+  getContentOfResultTemplate(
+    result: Result | FoldedResult
+  ): HTMLElement | DocumentFragment;
+}
+
+export interface ResultsProps
+  extends Pick<
+    AtomicResultListBase,
+    | 'display'
+    | 'density'
+    | 'host'
+    | 'resultListState'
+    | 'bindings'
+    | 'imageSize'
+    | 'image'
+    | 'getContentOfResultTemplate'
+  > {
+  classes: string;
+}
+
+export function getClasses({
+  display = 'list',
+  density,
+  imageSize,
+  image,
+  resultListState,
+  resultList,
+}: AtomicResultListBase): string {
+  const classes = getResultDisplayClasses(display, density, imageSize ?? image);
+  if (resultListState.firstSearchExecuted && resultList.state.isLoading) {
+    classes.push('loading');
+  }
+  return classes.join(' ');
+}
+
+export function getId(
+  result: Result | FoldedCollection,
+  resultListState: FoldedResultListState | ResultListState
+) {
+  return unfolded(result).uniqueId + resultListState.searchResponseId;
+}
+
+export function unfolded(result: Result | FoldedResult): Result {
+  return (result as FoldedResult).result || result;
+}
+
+const defaultFieldsToInclude = [
+  'date',
+  'author',
+  'source',
+  'language',
+  'filetype',
+  'parents',
+  'ec_price',
+  'ec_name',
+  'ec_description',
+  'ec_brand',
+  'ec_category',
+  'ec_item_group_id',
+  'ec_shortdesc',
+  'ec_thumbnails',
+  'ec_images',
+  'ec_promo_price',
+  'ec_in_stock',
+  'ec_cogs',
+  'ec_rating',
+];
+
+export function getFields(fieldsToInclude: string): string[] {
+  if (fieldsToInclude.trim() === '') return [...defaultFieldsToInclude];
+  return defaultFieldsToInclude.concat(
+    fieldsToInclude.split(',').map((field) => field.trim())
+  );
+}
+export function handleInfiniteScroll(
+  isEnabled: boolean,
+  host: HTMLElement,
+  resultList: ResultList | FoldedResultList
+) {
+  if (!isEnabled) {
+    return;
+  }
+
+  const hasReachedEndOfElement =
+    window.innerHeight + window.scrollY >= host.offsetHeight;
+
+  if (hasReachedEndOfElement) {
+    resultList.fetchMoreResults();
+  }
+}
+
+export async function registerResultTemplates(
```

Comment:
I'm not a huge fan of these functions (registerResultTemplates, initializeResultList, postRenderCleanup), which receives a reference to the component and then "augment it" in a non explicit way.

Said differently, I think we will have to jump around quite a bit in the code, across different files, to follow what's going on.

I think these could be refactored in a single class (lets say `ResultListCommon`). That class itself would contain the properties it needs to do its job (ie: the headless controller).

Then, in `atomic-result-list` we could do `this.resultListCommon = new ResultListCommon({...whatever is needed...})` when initializing.

Afterwards, let's say that `atomic-result-list` needs access to the resultsPerPage headless controller, it could do `this.resultListCommon.resultsPerPage`.

I think with that pattern we'd be able to look at `ResultListCommon` as one entity which contains everything it needs to do its job properly.

Reply:
ok, that makes sense! I had a feeling that this wasn't great, but I'm not really used to working with classes, so didn't have better ideas! I'll try to implement your suggestion